### ISSUE TITLE
EvalResult can't log 2D tensor

### ISSUE THREAD

**Issue by xiadingZ** [PyTorchLightning/pytorch-lightning#3697]
I want to calculate confusion matrix in test step, but EvalResult can't log it.  How can I do to store it in EvalResult? If I have to override `test_epoch_end`, what should I do to process logged metric and my confusion matrix as default behaviour? 
```
    def test_step(self, batch, batch_idx):
        result = pl.EvalResult()
        result.log('test_loss', loss, sync_dist=True)
        result.log('test_vacc', vid_acc, sync_dist=True)
        #result.log('confusionmatrix', cm, sync_dist=True, sync_dist_op=torch.distributed.ReduceOp.SUM)
        return result
```

**Comment by LucaButera** [comment]
If you are using classic TensorBoard logging you will not be able to log a non scalar tensor, as far as i know. 
The workaround I used for this is to store the confusion matrix at each ```test_step``` and then compute the average in the ```test_epoch_end```. After this you can convert the matrix into an image (this is the only way i found to log it to tensorboard. If you use a logger that can log 2D tensors do as you would). Finally I delete the confusion matrix from the result, in order to avoid errors.

Below the pseudo-code

```
def test_step(self, batch, batch_idx, *args, **kwargs):
    # code that leads to conf matrix
    confusion_matrix = # whatever you do to compute the confusion matrix
    result = EvalResult()
    result.confusion_matrix = confusion_matrix
    # whatever you want to log additionally
    return result

def test_epoch_end(self, outputs):
    confusion_matrix = torch.mean(outputs.confusion_matrix, dim=0)
    # whatever you need to do to log the matrix 
    # for example I use a function to compute the image as a tensor and then I add the image to the logger
    del outputs['confusion_matrix']
    # whatever aggregation you need to perform on other metrics
    return outputs
```

Mind that, during the ```test_step``` you may need to unsqueeze the confusion matrix tensor in order to make it of shape ```[1, n_classes, n_classes]``` instead of ```[n_classes, n_classes]```. This way (assuming they all have the same shape), they will be properly stacked into a tensor of shape ```[n_batches, n_classes, n_classes]``` instead of ```[n_batches * n_classes, n_classes]``` for the ```test_epoch_end```.

Hope this is of some help.

**Comment by williamFalcon** [comment]
For anything beyond scalars, you need to use the logger directly:

```python
tensorboard = self.logger.experiment.

tensorbaord.do_whatever_it_supports()
```

**Comment by xiadingZ** [comment]
> If you are using classic TensorBoard logging you will not be able to log a non scalar tensor, as far as i know.
> The workaround I used for this is to store the confusion matrix at each `test_step` and then compute the average in the `test_epoch_end`. After this you can convert the matrix into an image (this is the only way i found to log it to tensorboard. If you use a logger that can log 2D tensors do as you would). Finally I delete the confusion matrix from the result, in order to avoid errors.
> 
> Below the pseudo-code
> 
> ```
> def test_step(self, batch, batch_idx, *args, **kwargs):
>     # code that leads to conf matrix
>     confusion_matrix = # whatever you do to compute the confusion matrix
>     result = EvalResult()
>     result.confusion_matrix = confusion_matrix
>     # whatever you want to log additionally
>     return result
> 
> def test_epoch_end(self, outputs):
>     confusion_matrix = torch.mean(outputs.confusion_matrix, dim=0)
>     # whatever you need to do to log the matrix 
>     # for example I use a function to compute the image as a tensor and then I add the image to the logger
>     del outputs['confusion_matrix']
>     # whatever aggregation you need to perform on other metrics
>     return outputs
> ```
> 
> Mind that, during the `test_step` you may need to unsqueeze the confusion matrix tensor in order to make it of shape `[1, n_classes, n_classes]` instead of `[n_classes, n_classes]`. This way (assuming they all have the same shape), they will be properly stacked into a tensor of shape `[n_batches, n_classes, n_classes]` instead of `[n_batches * n_classes, n_classes]` for the `test_epoch_end`.
> 
> Hope this is of some help.

confusion_matrix returned in this way will automatically be synced? Or I need to do manually in `test_epoch_end`?

**Comment by LucaButera** [comment]
@xiadingZ The confusion matrix will not be automatically synced. As you can see I delete it from ```outputs``` as leaving it would result in an exception, since ```outputs.confusion_matrix``` is not a scalar. You should log your confusion matrix manually. For example I build an image of the confusion matrix and then I log it to tensorboard by doing

```
self.logger.experiment.add_image('confusion_matrix', confusion_matrix_image, self.global_step)
```

inside ```test_epoch_end```. 

To produce the image I use ```matplotlib```, then I save it in memory and I load it with ```Pillow```. Finally I convert it to a ```torch.Tensor```.

Mind that implementing ```test_epoch_end``` overrides the automatic aggregation on metrics performed by ```pytorch-lightning```. Hence you need to manually aggregate your other metrics (accuracy, loss and so on). At least this is true for version 0.9.0
